### ISSUE TITLE
Installation problems - setup: The program 'eta' is required but it could not be found.

### ISSUE THREAD

**Issue by ogkloo** [typelead/eta#238]
I'm trying to install Eta from source, and after chasing down a lot of dependencies I ran into this:

https://gist.github.com/ogkloo/66784cdfa2f8181195e7010b6dfd02ff

Trying to install on ubuntu 16.04, stack version 1.4.0.

**Comment by rahulmutt** [comment]
Try running `stack setup --allow-different-user` and then `stack install --allow-different-user`.

**Comment by ogkloo** [comment]
Setup runs successfully as previously. Install fails with the same error message.

**Comment by rahulmutt** [comment]
Can you run the following commands in the terminal and share the output?

```
echo $USER
ls -al ~/.stack
```

It seems you do not have permissions to modify `~/.stack`. By chance was stack installed on your user by a different user?

**Comment by ogkloo** [comment]
$USER: fir

total 48
drwxr-xr-x 11 fir  root 4096 Mar 16 11:37 .
drwxr-xr-x 22 fir  fir  4096 Mar 16 14:51 ..
drwxr-xr-x  2 root root 4096 Mar 16 01:10 build-plan
drwxr-xr-x  3 root root 4096 Mar 16 00:47 build-plan-cache
-rw-r--r--  1 root root  240 Mar 16 00:47 config.yaml
drwxr-xr-x  3 root root 4096 Mar 16 01:13 global-project
drwxr-xr-x  3 root root 4096 Mar 16 00:47 indices
drwxr-xr-x  4 root root 4096 Mar 16 14:52 precompiled
drwxr-xr-x  3 root root 4096 Mar 16 01:10 programs
drwxr-xr-x  3 root root 4096 Mar 16 00:48 setup-exe-cache
drwxrwxr-x  2 fir  fir  4096 Mar 16 11:37 setup-exe-src
drwxr-xr-x  3 root root 4096 Mar 16 00:48 snapshots

**Comment by rahulmutt** [comment]
Hmm that's a problem. It looks like all the key stack folders were given root permissions. Can you try

```sudo chmod -R fir:fir ~/.stack```

and run `./cleaninstall.sh` afterward?

**Comment by ogkloo** [comment]
Got rid of the root permissions, chmoded them to 777

ls -al ~/.stack:
```
total 48
drwxr-xr-x 11 777 fir 4096 Mar 16 11:37 .
drwxr-xr-x 22 fir fir 4096 Mar 16 14:51 ..
drwxr-xr-x  2 777 fir 4096 Mar 16 01:10 build-plan
drwxr-xr-x  3 777 fir 4096 Mar 16 00:47 build-plan-cache
-rw-r--r--  1 777 fir  240 Mar 16 00:47 config.yaml
drwxr-xr-x  3 777 fir 4096 Mar 16 01:13 global-project
drwxr-xr-x  3 777 fir 4096 Mar 16 00:47 indices
drwxr-xr-x  4 777 fir 4096 Mar 16 14:52 precompiled
drwxr-xr-x  3 777 fir 4096 Mar 16 01:10 programs
drwxr-xr-x  3 777 fir 4096 Mar 16 00:48 setup-exe-cache
drwxrwxr-x  2 777 fir 4096 Mar 16 11:37 setup-exe-src
drwxr-xr-x  3 777 fir 4096 Mar 16 00:48 snapshots
```

Also in cleaninstall.sh, I added --allow-different-user because otherwise it just quits immiediately.

./cleaninstall.sh:
```
You are not the owner of '/home/fir/.stack/'. Aborting to protect file permissions.
Retry with '--allow-different-user' to disable this precaution.
You are not the owner of '/home/fir/.stack/'. Aborting to protect file permissions.
Retry with '--allow-different-user' to disable this precaution.
ghc-paths-0.1.0.9: configure
Progress: 1/2
--  While building package ghc-paths-0.1.0.9 using:
      /tmp/stack4964/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/setup --builddir=.stack-work/dist/x86_64-linux/Cabal-1.22.5.0 configure --with-ghc=/home/fir/.stack/programs/x86_64-linux/ghc-7.10.3/bin/ghc --with-ghc-pkg=/home/fir/.stack/programs/x86_64-linux/ghc-7.10.3/bin/ghc-pkg --user --package-db=clear --package-db=global --package-db=/home/fir/.stack/snapshots/x86_64-linux/lts-6.27/7.10.3/pkgdb --package-db=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/pkgdb --libdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/lib --bindir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/bin --datadir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/share --libexecdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/libexec --sysconfdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/etc --docdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/doc/ghc-paths-0.1.0.9 --htmldir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/doc/ghc-paths-0.1.0.9 --haddockdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/doc/ghc-paths-0.1.0.9 --dependency=Cabal=Cabal-1.22.8.0-ba8852a979ca7ac5dbfeb9a926f4d208 --dependency=base=base-4.8.2.0-0d6d1084fbc041e1cded9228e80e264d --dependency=directory=directory-1.2.2.0-660a7a83a753ed85c8a374c15dae2b97
    Process exited with code: ExitFailure 1
    Logs have been written to: /home/fir/eta/.stack-work/logs/ghc-paths-0.1.0.9.log

    [1 of 2] Compiling Main             ( /tmp/stack4964/ghc-paths-0.1.0.9/Setup.hs, /tmp/stack4964/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/Main.o )
    [2 of 2] Compiling StackSetupShim   ( /home/fir/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /tmp/stack4964/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/StackSetupShim.o )
    Linking /tmp/stack4964/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/setup ...
    Configuring ghc-paths-0.1.0.9...
    setup: The program 'eta' is required but it could not be found.
```

stack install --allow-different-user:
```
ghc-paths-0.1.0.9: configure
Progress: 1/2
--  While building package ghc-paths-0.1.0.9 using:
      /tmp/stack5029/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/setup --builddir=.stack-work/dist/x86_64-linux/Cabal-1.22.5.0 configure --with-ghc=/home/fir/.stack/programs/x86_64-linux/ghc-7.10.3/bin/ghc --with-ghc-pkg=/home/fir/.stack/programs/x86_64-linux/ghc-7.10.3/bin/ghc-pkg --user --package-db=clear --package-db=global --package-db=/home/fir/.stack/snapshots/x86_64-linux/lts-6.27/7.10.3/pkgdb --package-db=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/pkgdb --libdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/lib --bindir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/bin --datadir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/share --libexecdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/libexec --sysconfdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/etc --docdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/doc/ghc-paths-0.1.0.9 --htmldir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/doc/ghc-paths-0.1.0.9 --haddockdir=/home/fir/eta/.stack-work/install/x86_64-linux/lts-6.27/7.10.3/doc/ghc-paths-0.1.0.9 --dependency=Cabal=Cabal-1.22.8.0-ba8852a979ca7ac5dbfeb9a926f4d208 --dependency=base=base-4.8.2.0-0d6d1084fbc041e1cded9228e80e264d --dependency=directory=directory-1.2.2.0-660a7a83a753ed85c8a374c15dae2b97
    Process exited with code: ExitFailure 1
    Logs have been written to: /home/fir/eta/.stack-work/logs/ghc-paths-0.1.0.9.log

    [1 of 2] Compiling Main             ( /tmp/stack5029/ghc-paths-0.1.0.9/Setup.hs, /tmp/stack5029/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/Main.o )
    [2 of 2] Compiling StackSetupShim   ( /home/fir/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /tmp/stack5029/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/StackSetupShim.o )
    Linking /tmp/stack5029/ghc-paths-0.1.0.9/.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/setup/setup ...
    Configuring ghc-paths-0.1.0.9...
    setup: The program 'eta' is required but it could not be found.
```

**Comment by rahulmutt** [comment]
Hmm, what are the permission for the `stack` executable itself and the `.stack` folder itself (not the contents)?

**Comment by ogkloo** [comment]
.stack folder:
`drwxr-xr-x 11  777 fir      4096 Mar 16 11:37 .stack`

stack executable (at ~/.local/stack, assuming that's the correct one):
`-rwxr-xr-x 1 root root 54448232 Mar 16 10:51 stack*`

**Comment by rahulmutt** [comment]
Hmm I'm not sure what the problem is. Can you try filing an issue on the `stack` repo since it seems mostly like a stack issue? 

https://github.com/commercialhaskell/stack/issues

**Comment by ogkloo** [comment]
Will do. Thank you for all the help.

**Comment by rahulmutt** [comment]
@ogkloo It seems to be fixed now, see #246

**Comment by rahulmutt** [comment]
Closing, as this should mostly be fixed.
